main.go

```diff
@@ -61,34 +65,109 @@ func printVersion() {
+		Port:                  0, // Add flag with default of 9443 when adding webhooks
+		Namespace:             watchNamespace,
+		ClientDisableCacheFor: []client.Object{&corev1.Secret{}},
+	})
+	if err != nil {
+		setupLog.Error(err, "unable to start manager")
+		return err
+	}
+
+	imgReconciler := metal3iocontroller.PreprovisioningImageReconciler{
+		Client:       mgr.GetClient(),
+		Log:          ctrl.Log.WithName("controllers").WithName("PreprovisioningImage"),
+		APIReader:    mgr.GetAPIReader(),
+		Scheme:       mgr.GetScheme(),
+		ImageHandler: imageServer,
+	}
+	if err = (&imgReconciler).SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", "PreprovisioningImage")
+		return err
+	}
+
+	// +kubebuilder:scaffold:builder
+
+	if err := setupChecks(mgr); err != nil {
+		return err
+	}
+
+	setupLog.Info("starting manager")
+	return mgr.Start(ctrl.SetupSignalHandler())
+}
+
+func loadStaticNMState(nmstateDir string, imageServer imagehandler.ImageHandler) error {
+	files, err := ioutil.ReadDir(nmstateDir)
+	if err != nil {
+		return errors.WithMessagef(err, "problem reading %s", nmstateDir)
+	}
+
+	for _, f := range files {
+		if f.IsDir() {
+			continue
+		}
+		b, err := os.ReadFile(path.Join(nmstateDir, f.Name()))
+		if err != nil {
+			return errors.WithMessagef(err, "problem reading %s", path.Join(nmstateDir, f.Name()))
+		}
+		igBuilder := ignition.New(b,
+			os.Getenv("IRONIC_BASE_URL"),
+			os.Getenv("IRONIC_AGENT_IMAGE"),
+			os.Getenv("IRONIC_AGENT_PULL_SECRET"),
+			os.Getenv("IRONIC_RAMDISK_SSH_KEY"),
+		)
+		ign, err := igBuilder.Generate()
+		if err != nil {
+			return errors.WithMessagef(err, "problem generating ignition %s", f.Name())
+		}
+		imageName := strings.Replace(f.Name(), ".yaml", ".qcow", 1) // master-1.yaml -> master-1.qcow
+		url, err := imageServer.ServeImage(imageName, ign)
+		if err != nil {
+			return err
+		}
+		setupLog.Info("serving image %s at %s", imageName, url)
+	}
+	return nil
 }
 
 func main() {
 	var watchNamespace string
 	var devLogging bool
 	var imagesBindAddr string
 	var imagesPublishAddr string
+	var startController bool
+	var nmstateDir string
 
 	// From CAPI point of view, BMO should be able to watch all namespaces
 	// in case of a deployment that is not multi-tenant. If the deployment
 	// is for multi-tenancy, then the BMO should watch only the provided
 	// namespace.
 	flag.StringVar(&watchNamespace, "namespace", os.Getenv("WATCH_NAMESPACE"),
-		"Namespace that the controller watches to reconcile host resources.")
+		"Namespace that the controller watches to reconcile preprovisioningimage resources.")
 	flag.StringVar(&imagesBindAddr, "images-bind-addr", ":8084",
 		"The address the images endpoint binds to.")
 	flag.StringVar(&imagesPublishAddr, "images-publish-addr", "http://127.0.0.1:8084",
 		"The address clients would access the images endpoint from.")
+	flag.BoolVar(&startController, "start-controller", true,
```

Comment:
Wondering if it could be worth having two separate binaries instead of one plus option